src/apps/monero/protocol/signing/step_01_init_transaction.py

```diff
@@ -0,0 +1,242 @@
+        many_outputs=True,
+        hmacs=hmacs,
+        rsig_data=rsig_data,
+    )
+
+
+def get_primary_change_address(state: State):
+    """
+    Computes primary change address for the current account index
+    """
+    from trezor.messages.MoneroAccountPublicAddress import MoneroAccountPublicAddress
+
+    D, C = monero.generate_sub_address_keys(
+        state.creds.view_key_private, state.creds.spend_key_public, state.account_idx, 0
+    )
+    return MoneroAccountPublicAddress(
+        view_public_key=crypto.encodepoint(C), spend_public_key=crypto.encodepoint(D)
+    )
+
+
+def check_change(state: State, outputs):
+    """
+    Checks if the change address is among tx outputs and it is equal to our address.
+    """
+    from apps.monero.xmr.sub.addr import addr_eq, get_change_addr_idx
+
+    change_idx = get_change_addr_idx(outputs, state.output_change)
+
+    change_addr = state.change_address()
+    if change_addr is None:
+        state.mem_trace("No change" if __debug__ else None)
+        return
+
+    if change_idx is None and state.output_change.amount == 0 and len(outputs) == 2:
+        state.mem_trace("Sweep tsx" if __debug__ else None)
+        return  # sweep dummy tsx
+
+    found = False
+    for out in outputs:
+        if addr_eq(out.addr, change_addr):
+            found = True
+            break
+
+    if not found:
+        raise misc.TrezorChangeAddressError("Change address not found in outputs")
+
+    my_addr = get_primary_change_address(state)
+    if not addr_eq(my_addr, change_addr):
+        raise misc.TrezorChangeAddressError("Change address differs from ours")
+
+    return True
+
+
+def process_payment_id(state: State, tsx_data):
+    """
+    Payment id -> extra
+    """
+    if common.is_empty(tsx_data.payment_id):
+        return
+
+    from apps.monero.xmr.sub import tsx_helper
+    from trezor import utils
+
+    if len(tsx_data.payment_id) == 8:
+        view_key_pub_enc = tsx_helper.get_destination_view_key_pub(
+            tsx_data.outputs, state.change_address()
+        )
+        if view_key_pub_enc == crypto.NULL_KEY_ENC:
+            raise ValueError(
+                "Destinations have to have exactly one output to support encrypted payment ids"
+            )
+
+        view_key_pub = crypto.decodepoint(view_key_pub_enc)
+        payment_id_encr = tsx_helper.encrypt_payment_id(
+            tsx_data.payment_id, view_key_pub, state.tx_priv
+        )
+
+        extra_nonce = payment_id_encr
+        extra_prefix = 1
+
+    elif len(tsx_data.payment_id) == 32:
+        extra_nonce = tsx_data.payment_id
+        extra_prefix = 0
+
+    else:
+        raise ValueError("Payment ID size invalid")
+
+    lextra = len(extra_nonce)
+    if lextra >= 255:
+        raise ValueError("Nonce could be 255 bytes max")
+
+    extra_buff = bytearray(3 + lextra)
+    extra_buff[0] = 2
+    extra_buff[1] = lextra + 1
+    extra_buff[2] = extra_prefix
+    utils.memcpy(extra_buff, 3, extra_nonce, 0, lextra)
```

Comment:
Imho we can use simple `extra_buff[2:] = extra_nonce` here, because we are not slicing the `extra_nonce` (therefore no extra allocation). cc @jpochyla

Reply:
👍